compiler/types.nim

```diff
@@ -814,6 +815,8 @@ template ifFastObjectTypeCheckFailed(a, b: PType, body: untyped) =
     if tfFromGeneric in a.flags * b.flags and a.sym.id == b.sym.id:
       # ok, we need the expensive structural check
       body
+    elif tfFromGeneric in unionFlags and tfFinal notin unionFlags:
```

Comment:
I don't understand this. Firstly, why would the types be equal when one comes from a generic instantiation and the other does not. Secondly, what has ``tfFinal`` to do here?